llvm/lib/CodeGen/FinalizeISel.cpp

```diff
@@ -66,13 +67,19 @@ static std::pair<bool, bool> runImpl(MachineFunction &MF) {
         // The expansion may involve new basic blocks.
         if (NewMBB != MBB) {
           PreserveCFG = false;
-          MBB = NewMBB;
-          I = NewMBB->getIterator();
-          MBBI = NewMBB->begin();
-          MBBE = NewMBB->end();
+          InsertNewMBB = true;
+          break;
         }
       }
     }
+
+    ++I;
+
+    // If we insert a new MBB, we should re-scan to avoid missing some MI.
+    if (InsertNewMBB) {
+      I = MF.begin();
+      InsertNewMBB = false;
+    }
```

Comment:
Can we just handle missed MIs correctly instead of iterating from begin again?

Reply:
Would it insert more than 1 MBB? I am not sure how should we set the MF iterator to make it at the first inserted MBB's iteration.

Reply:
I'm not familiar with RISCV, IIUC the old behavior may skip the rest MIs in `I` if new blocks are inserted after `I`, and skipped MIs may need expansion. If so, can we check the next BB is the new BB to set iterators correctly?

Reply:
I change to continue scan current MBB.